File: maximum_depth_of_binary_tree.py
'''
Created on 2014-2-21

leetcode : Maximum Depth of Binary Tree

Problem:

Given a binary tree, find its maximum depth.

The maximum depth is the number of nodes along 
the longest path from the root node down to the 
farthest leaf node.


Solve :




@author: xqk
'''

# Definition for a  binary tree node

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    def maxDepth(self, root):
        #=======================================================================
        #my method 
        #root.val = 1
        # q = [root]
        # lastNode = None
        # while len(q) != 0 :
        #     fNode = q[0]
        #     if fNode.left != None:
        #         fNode.left.val = fNode.val + 1;
        #         q.append(fNode.left)
        #     if fNode.right != None:
        #         fNode.right.val = fNode.val + 1
        #         q.append(fNode.right)
        #     lastNode = q[len(q)-1]
        #     q = q[1:]
        # return lastNode.val
        #=======================================================================
    
        if root == None :
            return 0;
        l = self.maxDepth(root.left)
        r = self.maxDepth(root.right)
        
        return  l + 1 if l > r else r + 1
         

    
root = TreeNode(2)
left = TreeNode(2)
leftRight = TreeNode(1)

left.right = leftRight
root.left = left



test = Solution()
print test.maxDepth(root)

File: Cplusplus/Maximum_Depth_of_Binary_Tree.cpp
/*
Leetcode : Maximum_Depth_of_Binary_Tree.cpp

Given a binary tree, find its maximum depth.

The maximum depth is the number of nodes along the

longest path from the root node down to the farthest

leaf node.


*/

#include <iostream>
#include <vector>

using namespace std;

// Definition for binary tree
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    int maxDepth(TreeNode *root) {
        if(root == NULL)
        {
            return 0;
        }
        else
        {
            int left = maxDepth(root->left);
            int right = maxDepth(root->right);
            int maxD = left > right ? left+1:right+1;
            return maxD;
        }
    }
};

int main()
{
    Solution test = Solution();
    TreeNode *root = new TreeNode(1);
    TreeNode *node2 = new TreeNode(2);
    TreeNode *node3 = new TreeNode(3);
    node2->right = node3;
    root->left = node2;
    cout<<test.maxDepth(root)<<endl;
    return 0;
}
